import-automation/progress-dashboard-rest/app/resource/progress_log.py

```diff
@@ -0,0 +1,166 @@
+'/system_runs/<string:run_id>/logs', and '/logs/<string:log_id>'.
+"""
+
+import enum
+
+import flask_restful
+from flask_restful import reqparse
+
+from app import utils
+from app.service import validation
+from app.service import import_attempt_database
+from app.service import system_run_database
+from app.service import progress_log_database
+from app.model import import_attempt_model
+from app.model import system_run_model
+from app.model import progress_log_model
+
+_ATTEMPT = import_attempt_model.ImportAttemptModel
+_RUN = system_run_model.SystemRunModel
+_LOG = progress_log_model.ProgressLogModel
+
+
+# TODO(intrepiditee): Move to model
+class LogLevel(enum.Enum):
+    """Allowed log levels of a log.
+
+    The level of a log can only be one of these.
+    """
+    CRITICAL = 'critical'
+    ERROR = 'error'
+    WARNING = 'warning'
+    INFO = 'info'
+    DEBUG = 'debug'
+
+
+LOG_LEVELS = frozenset(level.value for level in LogLevel)
+
+
+class ProgressLog(flask_restful.Resource):
+    """Base class for a progress log resource.
+
+    Attributes:
+        client: datastore Client object used to communicate with Datastore
+        run_database: SystemRunDatabase object for querying and storing
+            system runs using the client
+        log_database: ProgressLogDatabase object for querying and storing
+            progress logs using the client
+        attempt_database: ImportAttemptDatabase object for querying and storing
+            import attempts using the client
+    """
+    parser = reqparse.RequestParser()
+    required_fields = [(_LOG.level,), (_LOG.message,)]
+    optional_fields = [(_LOG.time_logged,), (_LOG.run_id,), (_LOG.attempt_id,)]
+    utils.add_fields(parser, required_fields, required=True)
+    utils.add_fields(parser, optional_fields, required=False)
+
+    # TODO(intrepiditee): Change other resources to also accept an optional arg
+    def __init__(self, client=None):
+        """Constructs a ProgressLog.
+
+        Args:
+            client: datastore Client object used to communicate with Datastore
+        """
+        if not client:
+            client = utils.create_datastore_client()
+        self.client = client
+        self.run_database = system_run_database.SystemRunDatabase(
+            self.client)
+        self.log_database = progress_log_database.ProgressLogDatabase(
+            self.client)
+        self.attempt_database = import_attempt_database.ImportAttemptDatabase(
+            self.client)
+
+
+class ProgressLogByID(ProgressLog):
+    """API associated with the endpoint '/logs/<string:log_id>' for querying
+    a progress log by its log_id.
+
+    Attributes:
+        See ImportLog.
+    """
+    # TODO(intrepiditee): Use a helper for get
+    # TODO(intrepiditee): Use exception for request errors
+    def get(self, log_id):
+        """Queries the progress logs by its log_id.
+
+        Args:
+            log_id: ID of the progress log as a string
+
+        Returns:
```

Comment:
Would it be easier to return a list, error message, and error code?  If the code is zero, then the list is valid.  Otherwise, the list should be ignored.

Reply:
Tuples returned from resources have to conform to specific forms: https://flask.palletsprojects.com/en/1.1.x/quickstart/#about-responses